### PR TITLE
[master] [DOCS] document using slots for file contents/data

### LINKED ISSUE #65222: [FEATURE REQUEST] support for contents via function call in file state modules
**Is your feature request related to a problem? Please describe.**
Currently, the ways to insert ad-hoc data from a function call into a file are limited to Jinja templating mechanisms such as:

```
something:
  file.managed:
    - name: /tmp/stuff
    - contents: |
        {% set things = salt.vault.read_secret("secret/my/secret") -%}
        {{ things }}
```

or

```
something:
  file.managed:
    - name: /tmp/stuff
    - template: jinja
    - source: salt://template.jinja
```

```
# /srv/salt/template.jinja
{% set things = salt.vault.read_secret("secret/my/secret") -%}
{{ things }}
```

The first option is problematic because secrets could be shown in `show_sls` output. The second is a pain because you'd need different template files for each set of function calls and parameters in use... or some terrible macro setup.

**Describe the solution you'd like**
It would be great if file state functions such as `file.managed` and `file.serialize` supported an integrated way to insert output from an execution module function call into a file via a `contents_function` parameter.

```yaml
something:
  file.managed:
    - name: /tmp/stuff
    - contents_function:
        - fun: vault.get
          args:
            - secret/my/secret
```

**Describe alternatives you've considered**
Workarounds shown in problem description.

**Please Note**
If this feature request would be considered a substantial change or addition, this should go through a SEP process here https://github.com/saltstack/salt-enhancement-proposals, instead of a feature request.

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
Updates documentation to make it clear how to use slots to achieve using execution module calls for file contents or data without the use of jinja.

### What issues does this PR fix or reference?
Fixes: #65222 

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Review by twangboy — DISMISSED** [review]

**Comment by OrangeDog** [comment]
These are all quite bad examples, as there's no need to use slots in these cases.

**Comment by nicholasmhughes** [comment]
> These are all quite bad examples, as there's no need to use slots in these cases.

Come up with some better ones then.

**Review by barbaricyawps — DISMISSED** [review]
@nicholasmhughes , this looks good to me. One non-blocking comment is that we use the Google Developer's Style Guide, which says to use sentence case instead of title case for headings. See: https://developers.google.com/style/headings

**Comment by OrangeDog** [comment]
> Come up with some better ones then.

Stuff that might be changed incidentally during state runs.

`group.getent`, `alternatives.display`, `pkg.version`, `quota.report`.

File: changelog/65222.added.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Document ability to supply contents via slots function call for file.managed and file.serialize
```
Comment (Ch3LL) [inline changelog/65222.added.md:1]: Doc changes should not include a changelog file. Can you remove please?

**Review by Ch3LL — CHANGES_REQUESTED** [review]

**Comment by Ch3LL** [comment]
I think these are good examples that help demonstrate how to use slots, if anything it helps get users started. if you want to add more documentation examples @OrangeDog please create a follow up PR, we would appreciate it.

**Comment by OrangeDog** [comment]
It doesn't need _more_ examples. It needs examples that don't steer people into using slots when they shouldn't be. Examples that demonstrate the point of slots.

IMO, this whole PR is unnecessary, as "you can use them anywhere" already includes "you can use them as the contents parameter in a file.managed state". If you are going to add a bunch more, at least have them make sense.

One of the problems with Salt documentation is being overly-specific, so people don't see the general case.

**Comment by mdschmitt** [comment]
I tend to agree; @OrangeDog if these are the wrong way to use slots, a couple good ways (and why) would be welcome.  This particular PR isn't massive, it ought to be easy enough to tailor it the way you think would be most helpful.

**Review by twangboy — DISMISSED** [review]
I think these examples are fine.

**Comment by whytewolf** [comment]
So, I agree with @OrangeDog these examples are not great. and could lead to further issues later if people think slots are only for `file.*` as all of the examples including the new ones are only calling out file states.

how about something like

```
add_group_known_users:
  group.present:
    - name: known_users

add_user:
  user.present:
    - name: foobar
    - uid: 600
    - gid: __slot__:salt:group.info("known_users").gid
    - require:
      - group: add_group_known_users
```

this calls out the dynamic nature of slots. showing that while the known_users group might not be present before the states run we can still get the gid of the group during the state run.

**Review by dmurphy18 — DISMISSED** [review]

**Comment by nicholasmhughes** [comment]
@whytewolf , thank you for contributing an example. I added it to the docs.

File: doc/topics/slots/index.rst [hunk 2]
```diff
@@ -68,3 +68,109 @@ Here is an example of result parsing and appending:
...
+
+    add_group_known_users:
+      group.present:
+        - name: known_users
+
+    add_user:
+      user.present:
+        - name: foobar
+        - uid: 600
+        - gid: __slot__:salt:group.info("known_users").gid
```
Comment (OrangeDog) [inline doc/topics/slots/index.rst:95]: This will not work because the argument is quoted.

File: doc/topics/slots/index.rst [hunk 3]
```diff
@@ -68,3 +68,109 @@ Here is an example of result parsing and appending:
...
+The following Salt state uses the `test.echo` execution function to generate the
+text "hello world." This output is then used as the content of the file
+`/tmp/things.txt`:
+
+.. code-block:: yaml
+
+    content-from-slots:
+      file.managed:
+        - name: /tmp/things.txt
+        - contents: __slot__:salt:test.echo("hello world")
```
Comment (OrangeDog) [inline doc/topics/slots/index.rst:136]: Same problem here, and below.

**Review by OrangeDog — CHANGES_REQUESTED** [review]

File: doc/topics/slots/index.rst (hunk 2 above)
Comment (nicholasmhughes) [reply doc/topics/slots/index.rst:95]: works just fine.

File: doc/topics/slots/index.rst (hunk 3 above)
Comment (nicholasmhughes) [reply doc/topics/slots/index.rst:136]: same. still works.

File: doc/topics/slots/index.rst (hunk 2 above)
Comment (OrangeDog) [reply doc/topics/slots/index.rst:95]: Hmm. Every time I've tried using quotes with slots it doesn't work properly.

**Review by dwoz — APPROVED** [review]

**Pull request merged**